### PR TITLE
feat: persistent mutable state

### LINKED ISSUE #211: bug: "X messages unread" doesn't disappear sometimes

### PR DESCRIPTION
**Pull request by alexis-kochetov**
Related to #218 
Resolves #211

### CODE REVIEW

File: src/dotnet/UI.Blazor/Services/PersistentState.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            .Subscribe(OnUpdated);
+    }
+
+    private void OnUpdated((IState<T> State, StateEventKind EventKind) tuple)
+    {
+        var (_, eventKind) = tuple;
+        if ((eventKind & StateEventKind.Updated) == 0 || !HasValue)
+            return;
+
+        lock (_lock)
```
Comment (FrolYo) [inline src/dotnet/UI.Blazor/Services/PersistentState.cs:60]: not clear what problem lock solves here

File: src/dotnet/UI.Blazor/Services/PersistentState.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+public interface IPersistentState<T> : IMutableState<T>, IAsyncDisposable
+{ }
+public class PersistentState<T>: MutableState<T>, IPersistentState<T>
+{
+    private readonly object _lock = new ();
+    private readonly CancellationTokenSource _cts;
+    private IDisposable? _updatedObservableCompletion;
+    private Task _lastPersistTask;
+
+    protected Func<T, ISessionCommand> PersistCommandFactory { get; }
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Services/PersistentState.cs:14]: I would say this is kinda complicated. I rather use `string Key` here & persist any of such states via `IPersistentStateStore.Persist`, which is typically a scoped service relying on whatever makes sense to do that.

**Review by alexyakunin — CHANGES_REQUESTED** [review]

File: src/dotnet/Chat.UI.Blazor/Services/Activity/UnreadMessages.cs [hunk 3]
```diff
@@ -1,27 +1,31 @@
...
     {
+        _session = session;
         _chats = chats;
-        _readPositions = readPositions;
+        _chatId = chatId;
+        _lastReadEntryId = lastReadEntryId;
     }
 
-    public async Task<int?> GetUnreadMessageCount(Session session, string chatId, CancellationToken cancellationToken)
+    public async Task<int?> GetUnreadMessageCount(CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (alexis-kochetov) [reply src/dotnet/Chat.UI.Blazor/Services/Activity/UnreadMessages.cs:20]: Agree

File: src/dotnet/UI.Blazor/Services/PersistentState.cs (hunk 2 above)
Comment (alexis-kochetov) [reply src/dotnet/UI.Blazor/Services/PersistentState.cs:14]: I don't like the idea that everything as a key-value pair

File: src/dotnet/UI.Blazor/Services/PersistentState.cs (hunk 1 above)
Comment (alexis-kochetov) [reply src/dotnet/UI.Blazor/Services/PersistentState.cs:60]: I would like to rely on memory barriers that lock makes to ensure we are updating last persist task

File: src/dotnet/UI.Blazor/Services/PersistentStateReplica.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
+using System.Diagnostics.CodeAnalysis;
+using ActualChat.Pooling;
+using Stl.Conversion;
+
+namespace ActualChat.UI.Blazor.Services;
+
+public sealed class PersistentStateReplica<T>: IPersistentState<T>
```
[the remark this answers is not in the document]
Comment (alexis-kochetov) [reply src/dotnet/UI.Blazor/Services/PersistentStateReplica.cs:7]: Renamed to `PersistentStateLease`

File: src/dotnet/UI.Blazor/Services/PersistentState.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    protected override void Initialize(State<T>.Options options)
+    {
+        base.Initialize(options);
+        var updatedObservable = Observable.FromEvent<Action<IState<T>, StateEventKind>, (IState<T>, StateEventKind)>(
+            onNext => (st, kind) => onNext((st, kind)),
+            h => Updated += h,
+            h => Updated -= h);
+        _updatedObservableCompletion = updatedObservable
+            .Sample((options as Options)?.SampleInterval ?? TimeSpan.FromMilliseconds(1000))
```
[the remark this answers is not in the document]
Comment (alexis-kochetov) [reply src/dotnet/UI.Blazor/Services/PersistentState.cs:50]: done

File: src/dotnet/UI.Blazor/Services/PersistentState.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    protected internal PersistentState(
+        Options options,
+        Func<T, ISessionCommand> persistCommandFactory,
+        IServiceProvider services)
+        : base(options, services, true)
+    {
+        PersistCommandFactory = persistCommandFactory;
+        UICommandRunner = services.GetRequiredService<UICommandRunner>();
+        _cts = new CancellationTokenSource();
```
[the remark this answers is not in the document]
Comment (alexis-kochetov) [reply src/dotnet/UI.Blazor/Services/PersistentState.cs:32]: Fixed

File: src/dotnet/UI.Blazor/Services/PersistentState.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    private readonly CancellationTokenSource _cts;
+    private IDisposable? _updatedObservableCompletion;
+    private Task _lastPersistTask;
+
+    protected Func<T, ISessionCommand> PersistCommandFactory { get; }
+    protected UICommandRunner UICommandRunner { get; }
+
+    public new record Options : MutableState<T>.Options
+    {
+        public TimeSpan SampleInterval { get; init; } = TimeSpan.FromMilliseconds(1000);
```
[the remark this answers is not in the document]
Comment (alexis-kochetov) [reply src/dotnet/UI.Blazor/Services/PersistentState.cs:19]: Sure

**Pull request merged**